File: vhosts/blog/nuldoc-src/dom.ts
export type Text = {
  kind: "text";
  content: string;
  raw: false;
};

export type RawHTML = {
  kind: "text";
  content: string;
  raw: true;
};

export type Element = {
  kind: "element";
  name: string;
  attributes: Map<string, string>;
  children: Node[];
};

export type Node = Element | Text | RawHTML;

export function addClass(e: Element, klass: string) {
  const classes = e.attributes.get("class");
  if (classes === undefined) {
    e.attributes.set("class", klass);
  } else {
    const classList = classes.split(" ");
    classList.push(klass);
    classList.sort();
    e.attributes.set("class", classList.join(" "));
  }
}

export function findFirstChildElement(
  e: Element,
  name: string,
): Element | null {
  for (const c of e.children) {
    if (c.kind === "element" && c.name === name) {
      return c;
    }
  }
  return null;
}

export function findChildElements(e: Element, name: string): Element[] {
  const cs = [];
  for (const c of e.children) {
    if (c.kind === "element" && c.name === name) {
      cs.push(c);
    }
  }
  return cs;
}

export function removeChildElements(e: Element, name: string) {
  e.children = e.children.filter((c) =>
    c.kind !== "element" || c.name !== name
  );
}

export function innerText(e: Element): string {
  let t = "";
  forEachChild(e, (c) => {
    if (c.kind === "text") {
      t += c.content;
    }
  });
  return t;
}

export function forEachChild(e: Element, f: (n: Node) => void) {
  for (const c of e.children) {
    f(c);
  }
}

export function forEachChildRecursively(e: Element, f: (n: Node) => void) {
  const g = (c: Node) => {
    f(c);
    if (c.kind === "element") {
      forEachChild(c, g);
    }
  };
  forEachChild(e, g);
}

export function text(content: string): Text {
  return {
    kind: "text",
    content: content,
    raw: false,
  };
}

export function el(
  name: string,
  attrs: [string, string][],
  ...children: Node[]
): Element {
  return {
    kind: "element",
    name: name,
    attributes: new Map(attrs),
    children: children,
  };
}

File: services/nuldoc/lib/nuldoc/dom.rb
module Nuldoc
  Text = Data.define(:content) do
    def kind = :text
  end

  RawNode = Data.define(:content) do
    def kind = :raw
  end

  Element = Data.define(:name, :attributes, :children) do
    def kind = :element
  end

  module DOM
    module_function

    def add_class(element, klass)
      classes = element.attributes['class']
      if classes.nil?
        element.attributes['class'] = klass
      else
        class_list = classes.split
        class_list.push(klass)
        class_list.sort!
        element.attributes['class'] = class_list.join(' ')
      end
    end

    def find_first_child_element(element, name)
      element.children.find { |c| c.kind == :element && c.name == name }
    end

    def find_child_elements(element, name)
      element.children.select { |c| c.kind == :element && c.name == name }
    end

    def inner_text(element)
      t = +''
      for_each_child(element) do |c|
        t << c.content if c.kind == :text
      end
      t
    end

    def for_each_child(element, &)
      element.children.each(&)
    end

    def for_each_child_recursively(element)
      g = proc do |c|
        yield(c)
        for_each_child(c, &g) if c.kind == :element
      end
      for_each_child(element, &g)
    end

    def for_each_element_of_type(root, element_name)
      for_each_child_recursively(root) do |n|
        yield(n) if n.kind == :element && n.name == element_name
      end
    end

    def map_children_recursively(element, &)
      element.children.map! do |c|
        c = yield(c)
        map_children_recursively(c, &) if c.kind == :element
        c
      end
    end

    def map_element_of_type(root, element_name, &)
      map_children_recursively(root) do |n|
        if n.kind == :element && n.name == element_name
          yield(n)
        else
          n
        end
      end
    end

    def process_text_nodes_in_element(element)
      new_children = []
      element.children.each do |child|
        if child.kind == :text
          new_children.concat(yield(child.content))
        else
          new_children.push(child)
        end
      end
      element.children.replace(new_children)
    end
  end
end
